consensus-types/hdiff/state_diff.go

```diff
@@ -0,0 +1,2101 @@
+	type sszSizeUnmarshaler interface {
+		ssz.Unmarshaler
+		ssz.Marshaler
+		proto.Message
+	}
+	var header sszSizeUnmarshaler
+	switch ret.targetVersion {
+	case version.Bellatrix:
+		header = &enginev1.ExecutionPayloadHeader{}
+	case version.Capella:
+		header = &enginev1.ExecutionPayloadHeaderCapella{}
+	case version.Deneb, version.Electra, version.Fulu:
+		header = &enginev1.ExecutionPayloadHeaderDeneb{}
+	default:
+		return errors.Errorf("unknown target version %d", ret.targetVersion)
+	}
+	if len(*data) < headerLength {
+		return errors.Wrap(errDataSmall, "executionPayloadHeader")
+	}
+	if err := header.UnmarshalSSZ((*data)[:headerLength]); err != nil {
+		return errors.Wrap(err, "failed to unmarshal executionPayloadHeader")
+	}
+	var err error
+	ret.executionPayloadHeader, err = blocks.NewWrappedExecutionData(header)
+	if err != nil {
+		return err
+	}
+	*data = (*data)[headerLength:]
+	return nil
+}
+
+func (ret *stateDiff) readWithdrawalIndices(data *[]byte) error {
+	if len(*data) < 16 {
+		return errors.Wrap(errDataSmall, "withdrawalIndices")
+	}
+	ret.nextWithdrawalIndex = binary.LittleEndian.Uint64((*data)[:8])
+	ret.nextWithdrawalValidatorIndex = primitives.ValidatorIndex(binary.LittleEndian.Uint64((*data)[8:16]))
+	*data = (*data)[16:]
+	return nil
+}
+
+func (ret *stateDiff) readHistoricalSummaries(data *[]byte) error {
+	if len(*data) < 8 {
+		return errors.Wrap(errDataSmall, "historicalSummaries")
+	}
+	historicalSummariesLength := int(binary.LittleEndian.Uint64((*data)[:8])) // lint:ignore uintcast
+	if len(*data) < 8+historicalSummariesLength*fieldparams.RootLength*2 {
+		return errors.Wrap(errDataSmall, "historicalSummaries")
+	}
+	ret.historicalSummaries = make([]*ethpb.HistoricalSummary, historicalSummariesLength)
+	cursor := 8
+	for i := range historicalSummariesLength {
+		ret.historicalSummaries[i] = &ethpb.HistoricalSummary{
+			BlockSummaryRoot: slices.Clone((*data)[cursor : cursor+fieldparams.RootLength]),
+			StateSummaryRoot: slices.Clone((*data)[cursor+fieldparams.RootLength : cursor+2*fieldparams.RootLength]),
+		}
+		cursor += 2 * fieldparams.RootLength
+	}
+	*data = (*data)[cursor:]
+	return nil
+}
+
+func (ret *stateDiff) readElectraPendingIndices(data *[]byte) error {
+	if len(*data) < 8*6 {
+		return errors.Wrap(errDataSmall, "electraPendingIndices")
+	}
+	ret.depositRequestsStartIndex = binary.LittleEndian.Uint64((*data)[:8])
+	ret.depositBalanceToConsume = primitives.Gwei(binary.LittleEndian.Uint64((*data)[8:16]))
+	ret.exitBalanceToConsume = primitives.Gwei(binary.LittleEndian.Uint64((*data)[16:24]))
+	ret.earliestExitEpoch = primitives.Epoch(binary.LittleEndian.Uint64((*data)[24:32]))
+	ret.consolidationBalanceToConsume = primitives.Gwei(binary.LittleEndian.Uint64((*data)[32:40]))
+	ret.earliestConsolidationEpoch = primitives.Epoch(binary.LittleEndian.Uint64((*data)[40:48]))
+	*data = (*data)[48:]
+	return nil
+}
+
+func (ret *stateDiff) readPendingDeposits(data *[]byte) error {
+	if len(*data) < 16 {
```

Comment:
wouldn't it be better to use params.BeaconConfig().MaxPendingDepositsPerEpoch

Reply:
The 16 here is not the number of deposits, is the length of the 2 indices that are read.